src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+        /// <returns>
+        /// True, if arguments and parameters match perfectly.
+        /// False, otherwise.
+        /// </returns>
+        // TODO: May need an API to replace this function,
+        // link: https://github.com/dotnet/roslyn/issues/42149
+        private static bool IsArgumentListAndParameterListPerfectMatch(SemanticModel semanticModel, SeparatedSyntaxList<ArgumentSyntax> arguments,
+            ImmutableArray<IParameterSymbol> parameters, ArgumentSyntax targetArgument, CancellationToken cancellationToken, out int targetParamIndex)
+        {
+            targetParamIndex = -1; // return invalid index if it is not a perfect match
+
+            var matchedTypes = new bool[parameters.Length]; // default value is false
+            var paramsMatchedByArray = false; // the parameter with keyword params can be either matched by an array type or a variable number of arguments
+            var inOrder = true; // assume the arguments are in order in default
+
+            for (var i = 0; i < arguments.Count; i++)
+            {
+                // Parameter index cannot out of its range, #arguments is larger than #parameter only if the last parameter with keyword params
+                var parameterIndex = Math.Min(i, parameters.Length - 1);
+
+                // If the argument has a name, get the corresponding parameter index
+                var nameSyntax = arguments[i].NameColon?.Name;
+                if (nameSyntax != null)
+                {
+                    var name = nameSyntax.ToString();
+                    var found = false;
+                    for (var j = 0; j < parameters.Length; j++)
+                    {
+                        if (name.Equals(parameters[j].Name))
+                        {
+                            // Check if the argument is in order with parameters.
+                            // If the argument breaks the order, the rest arguments of matched functions must have names
+                            if (i != j)
+                                inOrder = false;
+                            parameterIndex = j;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found) return false;
+                }
+
+                // The argument is either in order with parameters, or have a matched name with parameters
+                var argType = semanticModel.GetTypeInfo(arguments[i].Expression, cancellationToken);
+                if (argType.Type != null && (inOrder || nameSyntax is object))
+                {
+                    // The type of argument must be convertible to the type of parameter
+                    if (!parameters[parameterIndex].IsParams
+                        && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                    {
+                        if (matchedTypes[parameterIndex]) return false;
```

Comment:
```suggestion
                        if (matchedTypes[parameterIndex])
                            return false;
```

Reply:
also, i don't understand why we're returning false here...   this code needs comments.